contrib/submit-protocol/src/App/ProtocolForm.tsx

```diff
@@ -0,0 +1,227 @@
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import React from "react";
+import { initializeIcons } from "office-ui-fabric-react/lib/Icons";
+import { Fabric } from "office-ui-fabric-react/lib/Fabric";
+import { Label } from "office-ui-fabric-react/lib/Label";
+import { TextField } from "office-ui-fabric-react/lib/TextField";
+import { Panel, PanelType } from "office-ui-fabric-react/lib/Panel";
+import { DefaultButton, PrimaryButton } from "office-ui-fabric-react/lib/Button";
+import crypto from "crypto";
+import MonacoEditor from "react-monaco-editor";
+import jsyaml from "js-yaml";
+
+interface IProtocolProps {
+  api: string;
+  user: string;
+  token: string;
+  source ?: {
+    jobName: string;
+    user: string;
+  };
+}
+
+interface IProtocolState {
+  jobName: string;
+  protocol: {[key: string]: string | object};
+  protocolYAML: string;
+  showEditor: boolean;
+}
+
+export default class ProtocolForm extends React.Component<IProtocolProps, IProtocolState> {
+  constructor(props: IProtocolProps) {
+    super(props);
+    initializeIcons();
+
+    this.state = {
+      jobName: "",
+      protocol: Object.create(null),
+      protocolYAML: "",
+      showEditor: false,
+    };
+
+    const source = this.props.source;
+    if (source && source.jobName && source.user) {
+      window.fetch(
+        `${this.props.api}/api/v1/user/${source.user}/jobs/${source.jobName}/config`
+      ).then((res) => {
+        return res.json();
+      }).then((body) => {
+        const protocol = jsyaml.safeLoad(body);
+        this.setState({ protocol });
+        this.setJobName(`${source.jobName}_clone_${crypto.randomBytes(4).toString("hex")}`);
+      }).catch((err) => {
+        window.alert(err.message);
+      });
+    }
+  }
+
+  public render() {
+    return (
+      <>
+        <Fabric>
+          <Panel
+            isOpen={this.state.showEditor}
+            isLightDismiss={true}
+            onDismiss={this.closeEditor}
+            type={PanelType.largeFixed}
+            headerText="Protocol YAML Editor"
+          >
+            <MonacoEditor
+              width={800}
+              height={800}
+              value={this.state.protocolYAML}
+              onChange={this.editProtocol}
+              language="yaml"
+              theme="vs-dark"
+              options={{ wordWrap: "on", readOnly: false }}
+            />
+            <div style={{ marginTop: "15px" }}>
+              <PrimaryButton text="Save" onClick={this.saveEditor} style={{ marginRight: "10px" }}/>
+              <DefaultButton text="Discard" onClick={this.discardEditor} />
+            </div>
+          </Panel>
+          <div className="content">
+            <form className="modal-dialog">
```

Comment:
I think these class names are from bootstrap. But it seems the plugin doesn't require bootstrap now.

Reply:
Right, the form uses some styles from webportal bootstrap css.

The goal of this PR is an end-to-end protocol submission pipeline. The form UI is just a prototype without experience design.

Reply:
I prefer using bootstrap css modules here.
We can't assume that bootstrap will be loaded in web portal in the future.

Reply:
Added.

Reply:
changed to `react-bootstrap`.

Reply:
updated to css modules.